src/event.rs

```diff
@@ -506,6 +510,14 @@ pub struct KeyboardInput {
     pub modifiers: ModifiersState,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum CompositionEvent {
+    CompositionStart(),
+    CompositionUpdate(String, usize),
+    CompositionEnd(),
```

Comment:
* `CompositionStart` should be a unit variant rather than an empty tuple variant.
* The cursor position field in `CompositionUpdate` appears to be a character-wise index. Unless there's a compelling reason not to, I think we should convert it to a byte-wise index, as that's more typical of Rust conventions.
* I think `CompositionEnd` should also contain the composition text, if possible.

Reply:
> CompositionStart should be a unit variant rather than an empty tuple variant.
Fixed this part.

> The cursor position field in CompositionUpdate appears to be a character-wise index. Unless there's a compelling reason not to, I think we should convert it to a byte-wise index, as that's more typical of Rust conventions.
I think the use case of cursor position is for rendering cursor in preedit string. In this senario, character-wise index will be more useful for users. 

> I think CompositionEnd should also contain the composition text, if possible.
 If we follow the Web specification way, it should be better to contain the text, but original XIM event doesn't contain the composition text. I'm not sure which is the better way.

Reply:
> I think the use case of cursor position is for rendering cursor in preedit string. In this senario, character-wise index will be more useful for users.

This is a fair point. My concern is that character-wise indexing presents a potential source of confusion: If an IME system were to generate text containing combining characters, would a character-wise index count such combining characters or treat them as part of the base character that they modify? (I apologize if I'm using confusing or incorrect terminology; I'm not entirely familiar with IME or Unicode concepts.)

> If we follow the Web specification way, it should be better to contain the text, but original XIM event doesn't contain the composition text. I'm not sure which is the better way.

I checked the [UI events spec](https://w3c.github.io/uievents/#event-type-compositionupdate) for clarification on this and my understanding is that a `CompositionUpdate` event with an empty buffer value is generated when the user has cancelled the IME input. So, I think that, if the IME has actually been completed, a `CompositionUpdate("", 0)` event should not be generated before `CompositionEnd`. In my testing, I am receiving an empty update event immediately before the end event, after either cancelling or completing input.

> If we follow the Web specification way, it should be better to contain the text, but original XIM event doesn't contain the composition text. I'm not sure which is the better way.

I think it would be best to adhere as closely to the web spec as we can. At least, that's my tendency here as I'm unfamiliar with the use case of IME APIs and so I'd prefer to follow the guidance of others. To that end, I think we may also want to add `String, usize` fields to the `CompositionStart` event, as the web spec indicates that some IME systems may provide this information in the start event.

Reply:
> My concern is that character-wise indexing presents a potential source of confusion
That's a good point. I'm not very familiar with IME or Unicode concept either, but character-wise index might not the best way. Byte-wise index might not be the best way, but I agree that it will be less confusion.

> I think it would be best to adhere as closely to the web spec as we can.
OK. I'll revise my change to follow the web spec as much as possible.